### PR TITLE
add WString support

### PR DESCRIPTION
**Pull request by dirk-thomas**
Connect to ros2/rosidl#352.

While waiting for ros2/rcl_interfaces#58 the patch (without tests) can already be reviewed.

### CODE REVIEW

File: rosidl_generator_py/resource/_msg_support.c.em [hunk 1]
```diff
@@ -309,6 +319,31 @@ nested_type = '__'.join(type_.namespaced_name())
...
+      if (!encoded_item) {
+        Py_DECREF(seq_field);
+        Py_DECREF(field);
+        return false;
+      }
+      char * buffer;
+      Py_ssize_t length;
+      int rc = PyBytes_AsStringAndSize(encoded_item, &buffer, &length);
+      if (rc) {
+        Py_DECREF(seq_field);
```
Comment (jacobperron) [inline rosidl_generator_py/resource/_msg_support.c.em]: We should decref `encoded_item` here.

File: rosidl_generator_py/resource/_msg_support.c.em [hunk 2]
```diff
@@ -309,6 +319,31 @@ nested_type = '__'.join(type_.namespaced_name())
...
+        Py_DECREF(field);
+        return false;
+      }
+      // use offset of 2 to skip BOM mark
+      bool succeeded = rosidl_generator_c__U16String__assignn_from_char(&dest[i], buffer + 2, length - 2);
+      if (!succeeded) {
+        Py_DECREF(seq_field);
```
Comment (jacobperron) [inline rosidl_generator_py/resource/_msg_support.c.em]: Decref `encoded_item` (or move decref to be before L334).

File: rosidl_generator_py/resource/_msg_support.c.em [hunk 3]
```diff
@@ -370,6 +405,28 @@ nested_type = '__'.join(type_.namespaced_name())
...
+      Py_DECREF(field);
+      return false;
+    }
+    // use offset of 2 to skip BOM mark
+    bool succeeded = rosidl_generator_c__U16String__assignn_from_char(&ros_message->@(member.name), buffer + 2, length - 2);
+    if (!succeeded) {
+      Py_DECREF(field);
+      return false;
+    }
+    Py_DECREF(encoded_field);
```
Comment (jacobperron) [inline rosidl_generator_py/resource/_msg_support.c.em]: We should ensure this is decref'd before or in the two preceding failure cases.

Comment (dirk-thomas) [reply rosidl_generator_py/resource/_msg_support.c.em]: Good catch: see 03a891edf95220079d16febdf9ad3175647f0958.

File: rosidl_generator_py/resource/_msg_support.c.em (hunk 1 above)
Comment (dirk-thomas) [reply rosidl_generator_py/resource/_msg_support.c.em]: See https://github.com/ros2/rosidl_python/pull/47#discussion_r280238235

File: rosidl_generator_py/resource/_msg_support.c.em (hunk 2 above)
Comment (dirk-thomas) [reply rosidl_generator_py/resource/_msg_support.c.em]: See https://github.com/ros2/rosidl_python/pull/47#discussion_r280238235

File: rosidl_generator_py/resource/_msg_support.c.em (hunk 3 above)
Comment (jacobperron) [reply rosidl_generator_py/resource/_msg_support.c.em]: You moved one of them, but this one should still be moved.

Comment (dirk-thomas) [reply rosidl_generator_py/resource/_msg_support.c.em]: And here it is: 1edd2365b6af17e46abdef5951aa154eb78fb7a8.

File: rosidl_generator_py/resource/_msg.py.em [hunk 4]
```diff
@@ -376,7 +376,7 @@ if member.name in dict(inspect.getmembers(builtins)).keys():
             from collections.abc import Set
             from collections import UserList
             from collections import UserString
-@[  elif isinstance(type_, AbstractString) and type_.has_maximum_size()]@
+@[  elif isinstance(type_, AbstractGenericString) and type_.has_maximum_size()]@
```
Comment (jacobperron) [inline rosidl_generator_py/resource/_msg.py.em]: What about the condition at L394 too?

Comment (dirk-thomas) [reply rosidl_generator_py/resource/_msg.py.em]: Very good catch. Yes, that condition definitely needs update to check the length of WStrings too: 7fc6f0d2896faf8dba62d1edce3d0244ca2c115c

**Pull request merged**